_posts/2016-07-11-value-of-automated-e2e-testing.md

```diff
@@ -0,0 +1,58 @@
+---
+layout: post
+author: riikkanen
+title: >
+  The Value of Automated e2e-testing in a software project
+excerpt: >
+  A peek into automated e2e-testing done in a software project. The first try with Wicket Page Tester. The second and more sophisticated version with JBehave and Selenium.
+tags:
+- Java
+- testing
+- e2e testing
+- JBehave
+- Wicket
+- Selenium
+- programming
+---
+
+The software project concerned started in the beginning of Autumn in 2010. I had the honor to work in the project from its beginning to June 2014 when my maternity leave began. The project was about to renew a 30 years old, technically outdated registry and everything attached to it, such as the user inteface and all the integrations. The project was a success, after almost three years of hard work, the application was taken into serious use without any major hardships. The project is still alive and kicking, even though, nowadays mostly maintenance, but also some feature development. 
+
+##The application
+
+The application is coded with Java. The actual user interface of the registry is developed as a web application using Apache Wicket, which is an Java web application framework. The data is stored in Oracle database and queried with direct SQL-queries, without any intermediate layer such as Hibernate. There are also numerous integrations bubbling under, powered by Mule. In the project I worked as a software developer both in back and front ends.
+
+The views in the user interface are designed in such a way that the user can see the main item and all the related sub-items related at the same time. The main item could have several sub-items listed in a table-like manner. The main item has certain managing tools and actions and, of course, all the sub-items have a bunch of tools and actions too. Most of the tools are implemented as icons with actions attached. The state and visibility of the tools depend about the status of the main item, but also from the context. Is this the main item which created the sub-item, the main item which edited them or a main item having nothing to do with the others? In addition to, some notification icons need to be shown in certain situations. However, after all we had a number of different icons which should be visible if and only if the sunspots were in the right position. 
+
+At least for me, the many ways to show data were also a burden. The functionality behind hiding and showing the icons respectively was very easy to break by sincere mistake when implementing some new feature or tweaking an existing one. Also, the number of different permutations were so many that it was nearly impossible to make full regression testing just by clicking the web site manually. I think that was the main reason we started to write automated end-to-end-tests in quite an early phase of the project.
```

Comment:
..the many ways of showing the data..?
the number of different permutations were so vast that it was nearly ..?
I tend to think this was the... || I feel this was the ?